src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaOauthSecurityConfig.java

```diff
@@ -170,65 +241,157 @@ public String encode(CharSequence rawPassword) {
-			logger.error("Failed to configure the authenticationEntryPoint on ResourceServerSecurityConfigurer.", e);
+		@Bean
+		public JWKSource<SecurityContext> jwkSource(RSAKey rsaKey) {
+			JWKSet jwkSet = new JWKSet(rsaKey);
+			return (jwkSelector, securityContext) -> jwkSelector.select(jwkSet);
+		}
+
+		@Bean
+		@SuppressWarnings("unused")
+		public JwtEncoder jwtEncoder(JWKSource<SecurityContext> jwkSource) {
+			return new NimbusJwtEncoder(jwkSource);
+		}
+
+		@Bean
+		public JwtDecoder jwtDecoder(JWKSource<SecurityContext> jwkSource) {
+			OAuth2TokenValidator<Jwt> defaultValidator = JwtValidators.createDefault();
+
+			OAuth2TokenValidator<Jwt> revocationValidator = jwt -> {
+				OAuth2Authorization authorization = this.authorizationService.findByToken(jwt.getTokenValue(),
+						OAuth2TokenType.ACCESS_TOKEN);
+
+				if (authorization != null && authorization.getAccessToken().isActive()) {
+					return OAuth2TokenValidatorResult.success();
+				} else {
+					OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.INVALID_TOKEN, "The token is revoked",
+							"https://tools.ietf.org/html/rfc6750#section-3.1");
+					return OAuth2TokenValidatorResult.failure(error);
+				}
+			};
+
+			OAuth2TokenValidator<Jwt> validator = new DelegatingOAuth2TokenValidator<>(defaultValidator,
+					revocationValidator);
+			NimbusJwtDecoder decoder = (NimbusJwtDecoder) OAuth2AuthorizationServerConfiguration.jwtDecoder(jwkSource);
+			decoder.setJwtValidator(validator);
+			return decoder;
+		}
+
+		private static RSAKey generateRsa(String rsaKeyId, String rsaPrivateKey, String rsaPublicKey) throws Exception {
+			InputStream publicKeyStream;
+			InputStream privateKeyStream;
+
+			if (rsaPrivateKey.startsWith("classpath")) {
+				privateKeyStream = JWKConfig.class.getClassLoader()
+						.getResourceAsStream(rsaPrivateKey.replace("classpath:/", ""));
+			} else {
+				privateKeyStream = new FileInputStream(new File(rsaPrivateKey));
+			}
+
+			if (rsaPublicKey.startsWith("classpath")) {
+				publicKeyStream = JWKConfig.class.getClassLoader()
+						.getResourceAsStream(rsaPublicKey.replace("classpath:/", ""));
+			} else {
+				publicKeyStream = new FileInputStream(new File(rsaPublicKey));
+			}
+
+			RSAPublicKey publicKey = getRSAPublicKeyFromPem(publicKeyStream);
+			RSAPrivateKey privateKey = getRSAPrivateKeyFromPem(privateKeyStream);
+			return new RSAKey.Builder(publicKey).privateKey(privateKey).keyID(rsaKeyId).build();
+		}
+
+		private static RSAPrivateKey getRSAPrivateKeyFromPem(InputStream is) throws Exception {
+			byte[] keyBytes = is.readAllBytes();
+
+			String key = new String(keyBytes);
+			String privateKeyPem = key.replace("-----BEGIN PRIVATE KEY-----", "")
+					.replaceAll(System.lineSeparator(), "")
+					.replace("-----END PRIVATE KEY-----", "");
+
+			byte[] encoded = Base64.getDecoder().decode(privateKeyPem);
+
+			KeyFactory keyFactory = KeyFactory.getInstance("RSA");
```

Comment:
Other key-generation methods are supported in Spring Security, maybe I should instead switch over to use a Java KeyStore which can be generated with `keytool` command which should be present anywhere java is installed and would not require `openssl`.

Reply:
That would make sense to me actually. Though I guess it depends on how difficult it would be to make the change. Since you would now need to know the location of the keystore instead of the files containing the keys.

Reply:
Refactored to use java key store in [bc31bfd](https://github.com/phac-nml/irida/pull/1339/commits/bc31bfd2da4bd29e9dd7ff503b9921c866ff6381)